File: src/cleansia_android/core/src/main/java/cz/cleansia/core/ui/components/CleansiaDropdown.kt
package cz.cleansia.core.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Check
import androidx.compose.material.icons.outlined.KeyboardArrowDown
import androidx.compose.material.icons.outlined.Search
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.unit.dp

/**
 * One option in a [CleansiaDropdown]. [id] is what the caller stores
 * and compares against `selectedId`; [label] is what we render.
 */
data class CleansiaDropdownOption(val id: String, val label: String)

/**
 * Outlined-text-field-shaped dropdown that opens a modal bottom sheet
 * for the picker. Two reasons we don't reuse `ExposedDropdownMenuBox`:
 *   1. Its anchor popup is M3-chromed (square menu, no theming hook
 *      for the picker surface) which clashes with the app's flat
 *      outlined cards everywhere else.
 *   2. The popup gets clipped on screens where it'd overflow — the
 *      sheet always has room.
 *
 * Matches [CleansiaTextField] visually: same 12dp corners, same border
 * colors, same float-label rhythm, so a dropdown sits next to a text
 * field as a sibling, not as an alien control.
 *
 * Pass `searchable = true` to surface a search box at the top of the
 * sheet — useful for country / language pickers; we filter case-
 * insensitive on the option's label.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CleansiaDropdown(
    selectedId: String?,
    options: List<CleansiaDropdownOption>,
    onSelected: (String) -> Unit,
    label: String,
    modifier: Modifier = Modifier,
    helper: String? = null,
    errorText: String? = null,
    placeholder: String? = null,
    enabled: Boolean = true,
    searchable: Boolean = false,
) {
    var sheetOpen by remember { mutableStateOf(false) }
    val selected = options.firstOrNull { it.id == selectedId }
    val isError = errorText != null

    DropdownAnchor(
        label = label,
        value = selected?.label,
        placeholder = placeholder,
        helper = helper,
        errorText = errorText,
        isError = isError,
        enabled = enabled,
        onClick = { if (enabled) sheetOpen = true },
        modifier = modifier,
    )

    if (sheetOpen) {
        val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
        ModalBottomSheet(
            onDismissRequest = { sheetOpen = false },
            sheetState = sheetState,
            containerColor = MaterialTheme.colorScheme.surface,
            dragHandle = null,
        ) {
            DropdownSheetBody(
                title = label,
                options = options,
                selectedId = selectedId,
                searchable = searchable,
                onSelect = { id ->
                    onSelected(id)
                    sheetOpen = false
                },
            )
        }
    }
}

/**
 * The visible row that mimics an outlined text field: outlined border,
 * float label, trailing chevron, optional helper / error supporting
 * text underneath. Click goes through to the sheet.
 */
@Composable
private fun DropdownAnchor(
    label: String,
    value: String?,
    placeholder: String?,
    helper: String?,
    errorText: String?,
    isError: Boolean,
    enabled: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val borderColor = when {
        isError -> MaterialTheme.colorScheme.error
        else -> MaterialTheme.colorScheme.outline
    }
    val labelColor = when {
        isError -> MaterialTheme.colorScheme.error
        !enabled -> MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.6f)
        else -> MaterialTheme.colorScheme.onSurfaceVariant
    }
    val valueColor = when {
        !enabled -> MaterialTheme.colorScheme.onSurfaceVariant
        value != null -> MaterialTheme.colorScheme.onSurface
        else -> MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.6f)
    }

    Column(modifier = modifier.fillMaxWidth()) {
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(12.dp))
                .border(1.dp, borderColor, RoundedCornerShape(12.dp))
                .clickable(enabled = enabled) { onClick() }
                .padding(horizontal = 16.dp, vertical = 10.dp),
        ) {
            Column {
                Text(
                    text = label,
                    style = MaterialTheme.typography.labelMedium,
                    color = labelColor,
                )
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text(
                        text = value ?: placeholder.orEmpty(),
                        style = MaterialTheme.typography.bodyLarge.copy(fontWeight = FontWeight.Medium),
                        color = valueColor,
                        modifier = Modifier.weight(1f),
                    )
                    Icon(
                        imageVector = Icons.Outlined.KeyboardArrowDown,
                        contentDescription = null,
                        tint = MaterialTheme.colorScheme.onSurfaceVariant,
                        modifier = Modifier.size(20.dp),
                    )
                }
            }
        }
        if (errorText != null || helper != null) {
            Text(
                text = errorText ?: helper.orEmpty(),
                style = MaterialTheme.typography.labelSmall,
                color = if (isError) MaterialTheme.colorScheme.error
                else MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.padding(start = 16.dp, top = 4.dp),
            )
        }
    }
}

@Composable
private fun DropdownSheetBody(
    title: String,
    options: List<CleansiaDropdownOption>,
    selectedId: String?,
    searchable: Boolean,
    onSelect: (String) -> Unit,
) {
    var query by remember { mutableStateOf("") }
    val filtered = remember(query, options) {
        if (!searchable || query.isBlank()) options
        else options.filter { it.label.contains(query.trim(), ignoreCase = true) }
    }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .heightIn(max = 560.dp),
    ) {
        if (title.isNotBlank()) {
            Text(
                text = title,
                style = MaterialTheme.typography.titleLarge.copy(fontWeight = FontWeight.Bold),
                color = MaterialTheme.colorScheme.onSurface,
                modifier = Modifier.padding(horizontal = 20.dp, vertical = 16.dp),
            )
        }
        if (searchable) {
            DropdownSearchField(value = query, onValueChange = { query = it })
            Spacer(Modifier.height(8.dp))
        }
        HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant.copy(alpha = 0.5f))
        LazyColumn(
            modifier = Modifier.fillMaxWidth(),
            contentPadding = androidx.compose.foundation.layout.PaddingValues(
                bottom = 12.dp,
            ),
        ) {
            items(items = filtered, key = { it.id }) { option ->
                DropdownOptionRow(
                    option = option,
                    isSelected = option.id == selectedId,
                    onClick = { onSelect(option.id) },
                )
            }
        }
    }
}

@Composable
private fun DropdownSearchField(value: String, onValueChange: (String) -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 12.dp, vertical = 8.dp)
            .clip(RoundedCornerShape(10.dp))
            .background(MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.6f))
            .padding(horizontal = 12.dp, vertical = 10.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Icon(
            imageVector = Icons.Outlined.Search,
            contentDescription = null,
            tint = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.size(18.dp),
        )
        Spacer(Modifier.width(8.dp))
        BasicTextField(
            value = value,
            onValueChange = onValueChange,
            singleLine = true,
            textStyle = TextStyle(
                color = MaterialTheme.colorScheme.onSurface,
                fontSize = MaterialTheme.typography.bodyMedium.fontSize,
            ),
            cursorBrush = SolidColor(MaterialTheme.colorScheme.primary),
            keyboardOptions = androidx.compose.foundation.text.KeyboardOptions(imeAction = ImeAction.Search),
            modifier = Modifier.weight(1f),
            decorationBox = { inner ->
                if (value.isEmpty()) {
                    Text(
                        text = "Search",
                        style = MaterialTheme.typography.bodyMedium,
                        color = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.7f),
                    )
                }
                inner()
            },
        )
    }
}

@Composable
private fun DropdownOptionRow(
    option: CleansiaDropdownOption,
    isSelected: Boolean,
    onClick: () -> Unit,
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .padding(horizontal = 16.dp, vertical = 14.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            text = option.label,
            style = MaterialTheme.typography.bodyLarge.copy(
                fontWeight = if (isSelected) FontWeight.SemiBold else FontWeight.Normal,
            ),
            color = if (isSelected) MaterialTheme.colorScheme.primary
            else MaterialTheme.colorScheme.onSurface,
            modifier = Modifier.weight(1f),
        )
        if (isSelected) {
            Icon(
                imageVector = Icons.Outlined.Check,
                contentDescription = null,
                tint = MaterialTheme.colorScheme.primary,
                modifier = Modifier.size(20.dp),
            )
        }
    }
}

File: src/cleansia_ios/CleansiaCore/Sources/CleansiaCore/Components/CleansiaDropdown.swift
import SwiftUI

public struct CleansiaDropdownOption: Identifiable, Equatable {
    public let id: String
    public let label: String

    public init(id: String, label: String) {
        self.id = id
        self.label = label
    }
}

public struct CleansiaDropdown: View {
    @Binding private var selectedId: String?
    private let options: [CleansiaDropdownOption]
    private let label: String
    private let helper: String?
    private let errorText: String?
    private let placeholder: String?
    private let enabled: Bool
    private let searchable: Bool

    @State private var sheetOpen = false

    public init(
        selectedId: Binding<String?>,
        options: [CleansiaDropdownOption],
        label: String,
        helper: String? = nil,
        errorText: String? = nil,
        placeholder: String? = nil,
        enabled: Bool = true,
        searchable: Bool = false
    ) {
        _selectedId = selectedId
        self.options = options
        self.label = label
        self.helper = helper
        self.errorText = errorText
        self.placeholder = placeholder
        self.enabled = enabled
        self.searchable = searchable
    }

    private var isError: Bool { errorText != nil }
    private var selected: CleansiaDropdownOption? { options.first { $0.id == selectedId } }

    public var body: some View {
        VStack(alignment: .leading, spacing: Spacing.xxs) {
            Button {
                if enabled { sheetOpen = true }
            } label: {
                VStack(alignment: .leading, spacing: 2) {
                    Text(label)
                        .font(CleansiaTypography.labelMedium)
                        .foregroundColor(labelColor)
                    HStack {
                        Text(selected?.label ?? placeholder ?? "")
                            .font(CleansiaTypography.bodyLarge)
                            .foregroundColor(valueColor)
                            .frame(maxWidth: .infinity, alignment: .leading)
                        Image(systemName: "chevron.down")
                            .font(.system(size: 14))
                            .foregroundColor(CleansiaColors.onSurfaceVariant)
                    }
                }
                .padding(.horizontal, Spacing.m)
                .padding(.vertical, 10)
                .frame(maxWidth: .infinity)
                .overlay(
                    RoundedRectangle(cornerRadius: CornerRadius.small)
                        .stroke(isError ? CleansiaColors.error : CleansiaColors.outline, lineWidth: 1)
                )
            }
            .buttonStyle(.plain)
            .disabled(!enabled)

            if let supporting = errorText ?? helper {
                Text(supporting)
                    .font(CleansiaTypography.labelSmall)
                    .foregroundColor(isError ? CleansiaColors.error : CleansiaColors.onSurfaceVariant)
                    .padding(.horizontal, Spacing.m)
            }
        }
        .sheet(isPresented: $sheetOpen) {
            DropdownSheet(
                title: label,
                options: options,
                selectedId: selectedId,
                searchable: searchable,
                onSelect: { id in
                    selectedId = id
                    sheetOpen = false
                }
            )
            .presentationDetents([.medium, .large])
        }
    }

    private var labelColor: Color {
        if isError { return CleansiaColors.error }
        return enabled ? CleansiaColors.onSurfaceVariant : CleansiaColors.onSurfaceVariant.opacity(0.6)
    }

    private var valueColor: Color {
        if selected != nil { return CleansiaColors.onSurface }
        return CleansiaColors.onSurfaceVariant.opacity(0.6)
    }
}

private struct DropdownSheet: View {
    let title: String
    let options: [CleansiaDropdownOption]
    let selectedId: String?
    let searchable: Bool
    let onSelect: (String) -> Void

    @State private var query = ""

    private var filtered: [CleansiaDropdownOption] {
        guard searchable, !query.trimmingCharacters(in: .whitespaces).isEmpty else { return options }
        return options.filter { $0.label.range(of: query, options: .caseInsensitive) != nil }
    }

    var body: some View {
        NavigationStack {
            List(filtered) { option in
                Button {
                    onSelect(option.id)
                } label: {
                    HStack {
                        Text(option.label)
                            .font(CleansiaTypography.bodyLarge)
                            .fontWeight(option.id == selectedId ? .semibold : .regular)
                            .foregroundColor(option.id == selectedId ? CleansiaColors.primary : CleansiaColors.onSurface)
                        Spacer()
                        if option.id == selectedId {
                            Image(systemName: "checkmark").foregroundColor(CleansiaColors.primary)
                        }
                    }
                }
            }
            .listStyle(.plain)
            .navigationTitle(title)
            .navigationBarTitleDisplayMode(.inline)
            .modifier(SearchableIfNeeded(enabled: searchable, query: $query))
        }
    }
}

private struct SearchableIfNeeded: ViewModifier {
    let enabled: Bool
    @Binding var query: String

    func body(content: Content) -> some View {
        if enabled {
            content.searchable(text: $query)
        } else {
            content
        }
    }
}

#if DEBUG
struct CleansiaDropdown_Previews: PreviewProvider {
    static var previews: some View {
        StatefulPreviewWrapper(String?.some("cz")) { binding in
            CleansiaDropdown(
                selectedId: binding,
                options: [
                    .init(id: "cz", label: "Czechia"),
                    .init(id: "sk", label: "Slovakia"),
                ],
                label: "Country",
                searchable: true
            )
            .padding()
        }
    }
}
#endif
